src/Elastic.Apm/Metrics/MetricSet.cs

```diff
@@ -8,8 +9,8 @@ namespace Elastic.Apm.Metrics
 	[JsonConverter(typeof(MetricSetConverter))]
 	internal class MetricSet : IMetricSet
 	{
-		public MetricSet(long timeStamp, List<MetricSample> samples)
-			=> (TimeStamp, Samples) = (timeStamp, samples);
+		public MetricSet(long timeStamp, IEnumerable<MetricSample> samples)
```

Comment:
```suggestion
		public MetricSet(long timeStamp, IEnumerable<MetricSample> samples, IApmLogger logger)
		{
			var metricSamples = samples as MetricSample[] ?? samples.ToArray();
			(TimeStamp, Samples) = (timeStamp, metricSamples.Where(x => !double.IsNaN(x.KeyValue.Value) && !double.IsInfinity(x.KeyValue.Value)));
			if (Samples.Count() != metricSamples.Length)
			{
				logger.Debug()?
					.Log("Some values were filtered out from MetricSample due to IsNan && IsInfinity, key: {key}", metricSamples.First()?.KeyValue.Key);
			}
		}
```

Reply:
Sure, I also thought about logging, however, I don't sure that adding IApmLogger as a dependency as a good idea. Is it better to use `ApmAgent.Logger` for this?

Reply:
I agree that it's a little bit odd to pass the logger to `MetricSet`. On the other hand by doing filtering in `MetricSet` we made `MetricSet` into more than just a data transfer object, it has some filtering logic, so I think adding a logger to `MetricSet` is justified with that.

>  Is it better to use ApmAgent.Logger for this?

You mean `Agent.Instance.Logger`? I'd try to avoid using the static instance, because it causes all sorts of problems during testing - when possible I'd just pass an instance.

Overall I'd be fine to just pass a logger instance, another option could be to either just do the same check on `Count` [here](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Metrics/MetricsCollector.cs#L137) when we build the `MetricSet`, but that'd mean we have to remember to do this check in case we build `MetricSet`s at other places, or another approach would be to add a flag with `IsFiltered` to `MetricsSet` and check on it. I personally would just stick with this simple solution by passing a logger - or maybe even make it optional if we have too many usages without a logger.